File: packages/verifier-ts/src/hex.ts
// Hex helpers. Kept tiny and dependency-free so the package
// surface stays auditable.

export function hexToBytes(hex: string): Uint8Array {
  if (typeof hex !== "string") {
    throw new TypeError("hex must be a string");
  }
  if (hex.length % 2 !== 0) {
    throw new Error("hex must have even length");
  }
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    const byte = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    if (Number.isNaN(byte)) {
      throw new Error(`invalid hex character at offset ${i * 2}`);
    }
    out[i] = byte;
  }
  return out;
}

export function bytesToHex(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length; i++) {
    // `bytes[i]` is `number | undefined` under noUncheckedIndexedAccess;
    // we just bounded `i` against `.length` so it's defined.
    const b = bytes[i];
    if (b === undefined) {
      throw new Error("bytesToHex: out-of-range index (should be unreachable)");
    }
    out += b.toString(16).padStart(2, "0");
  }
  return out;
}

File: packages/client-py/src/uniclaw_client/_hex.py
"""Hex helpers. Kept tiny and dependency-free so the auditable surface
stays small."""

from __future__ import annotations


def hex_to_bytes(hex_str: str) -> bytes:
    """Decode a hex string into raw bytes.

    Raises:
        TypeError: if ``hex_str`` is not a ``str``.
        ValueError: if the length is odd or any character isn't hex.
    """
    if not isinstance(hex_str, str):
        raise TypeError("hex must be a str")
    if len(hex_str) % 2 != 0:
        raise ValueError("hex must have even length")
    try:
        return bytes.fromhex(hex_str)
    except ValueError as e:
        raise ValueError(f"invalid hex: {e}") from e


def bytes_to_hex(data: bytes) -> str:
    """Encode raw bytes as a lowercase hex string."""
    return data.hex()
